### PR TITLE
Fixed calltip generation to account for generic functions and CUDA ke…

### PR DESCRIPTION
**Pull request by DThought**
Languages such as C# and Java have generic functions of the form

    static void Swap<T>(ref T lhs, ref T rhs)
    {
        T temp;
        temp = lhs;
        lhs = rhs;
        rhs = temp;
    }

(taken from <https://msdn.microsoft.com/en-us/library/twcad0zb.aspx>). Although the tags from these files are parsed correctly, calltip generation fail to take into account the type argument (in angle brackets) in calls of the form `Swap<int>(ref a, ref b);`.

This commit allows the calltip to find the function name `Swap` so that the calltip is generated. It does not affect calls that already work, such as those of the form `Swap(ref a, ref b);`.

In addition, this allows .cu (CUDA) files to use the C tag parser, which until now would have failed on kernel calls of the form `MatAdd<<<numBlocks, threadsPerBlock>>>(A, B, C);` due to the grid parameters. It also makes it possible to support LaTeX calltips of the form `\mynewcommand[opt]{Hello People}` in the future.

### CODE REVIEW

**Comment by elextr** [comment]
How does this change the C tags parser?  You don't seem to touch it?

**Comment by DThought** [comment]
It doesn't change the parser at all—just meant that calltips in CUDA are useful now that the grid parameters can be ignored, so I added the C tag parser to the default CUDA code as you suggested [here](https://github.com/geany/geany/issues/494). Sorry for the confusion.

**Comment by elextr** [comment]
No problem, I read it that the parser would have failed and you had done something to fix it.

**Comment by elextr** [comment]
LGTM, not tested but should also work for C++ explicitly specialized function calls. @b4n?

File: src/editor.c [hunk 1]
```diff
@@ -1953,7 +1953,17 @@ gboolean editor_show_calltip(GeanyEditor *editor, gint pos)
...
 
 	word[0] = '\0';
+
+	if (utils_is_closing_brace(sci_get_char_at(sci, pos - 1), TRUE))
+	{
+		/* optional parameters here, so move backwords from current position to skip it */
+		pos = brace_match(sci, pos - 1, TRUE);
+		if (pos == -1)
+			return FALSE;
+	}
```
Comment (b4n) [inline src/editor.c]: This should rather be before `word[0] = '\0';`

File: src/editor.c [hunk 2]
```diff
@@ -1953,7 +1953,17 @@ gboolean editor_show_calltip(GeanyEditor *editor, gint pos)
 		return FALSE;
 
 	word[0] = '\0';
+
```
Comment (b4n) [inline src/editor.c]: you need to allow space(s) between the `<>` and the `(`, something like this:
```c
if (pos > 0 && isspace(sci_get_char_at(sci, pos - 1)))
	pos--;
```
(or even a `while` to allow any number of spaces, not only one, but we don't have it currently so probably not necessary)

Comment (b4n) [inline src/editor.c]: FTR, for code like
```C++
foo<bar> (hello);
```

File: src/utils.c [hunk 3]
```diff
@@ -157,33 +157,36 @@ gint utils_get_line_endings(const gchar* buffer, gsize size)
 
 gboolean utils_isbrace(gchar c, gboolean include_angles)
 {
+	return utils_is_opening_brace(c, include_angles)
+		|| utils_is_closing_brace(c, include_angles);
```
Comment (b4n) [inline src/utils.c]: `||` should go at the end of the previous line (yeah I know there are some exceptions to this in the current code, but lets not add more :)

**Comment by b4n** [comment]
Otherwise LGTM, but ideally the CUDA change should be in a separate commit

**Comment by DThought** [comment]
Committed the style changes you suggested.

File: src/editor.c [hunk 4]
```diff
@@ -1952,8 +1952,20 @@ gboolean editor_show_calltip(GeanyEditor *editor, gint pos)
...
+	{
+		/* optional parameters here, so move backwords from current position to skip it */
+		pos = brace_match(sci, pos - 1, TRUE);
+		if (pos == -1)
+			return FALSE;
+	}
+
 	word[0] = '\0';
 	editor_find_current_word(editor, pos - 1, word, sizeof word, NULL);
+
```
Comment (b4n) [inline src/editor.c]: this extra line shouldn't be there

**Comment by b4n** [comment]
code looks ok now.  @elextr as you seemed to like that, I'd love your testing too ;)

**Comment by b4n** [comment]
I committed a simpler version not altering brace_match(), which was actually not needed as `sci_find_matching_brace()` is just fine.  Also, for now only handle `<>` to avoid possible false-positives.

The CUDA filetype part was already fixed in 9365110c4c0fa23c99cb9b79149f68f4049e342e (#830, #831).

**Pull request closed without merging**